### PR TITLE
Add support for context overrides

### PR DESCRIPTION
**Pull request by iFrostizz**
A `TestContext` can now be passed to the simulator in order to override any value of the `callInfo`.
This is firstly used to call a program (alter the `ProgramId`), but also to spoof the timestamp, block height, actor, and so on.

Closes https://github.com/ava-labs/hypersdk/issues/1062

### CODE REVIEW

File: x/programs/cmd/simulator/src/lib.rs [hunk 1]
```diff
@@ -91,6 +93,27 @@ pub enum Key {
...
+pub struct TestContext {
+    pub program_id: Id,
+}
+
+#[derive(Clone, Debug, PartialEq, Serialize)]
+#[serde(tag = "type", rename = "testContext")]
+pub(crate) struct SimulatorTestContext {
+    #[serde(serialize_with = "base64_struct_encode")]
+    value: TestContext,
+}
```
Comment (iFrostizz) [inline x/programs/cmd/simulator/src/lib.rs:108]: It's clearly a hack. There is no way to externally tag a struct to have this `type` `value` structure like we have for enums.

File: x/programs/rust/wasmlanche-sdk/src/program.rs [hunk 2]
```diff
@@ -24,22 +25,14 @@ pub enum ExternalCallError {
 /// Represents the current Program in the context of the caller. Or an external
 /// program that is being invoked.
 #[cfg_attr(feature = "debug", derive(Debug))]
+#[derive(Clone, BorshSerialize, Serialize)]
```
Comment (iFrostizz) [inline x/programs/rust/wasmlanche-sdk/src/program.rs:28]: I will back off this `Serialize` add. I wanted to be able to pass the `Context` but it was not really easy to construct it in a nice way because of traits such as `Default` for `Program<K>`. Now that we have the `TestContext` we have a clean builder API.

File: x/programs/runtime/util_test.go [hunk 3]
```diff
@@ -20,12 +20,14 @@ type testRuntime struct {
 	DefaultGas uint64
 }
 
-func (t *testRuntime) CallProgram(program codec.Address, actor codec.Address, function string, params ...interface{}) ([]byte, error) {
+func (t *testRuntime) CallProgram(program codec.Address, actor codec.Address, height uint64, timestamp uint64, function string, params ...interface{}) ([]byte, error) {
```
Comment (dboehm-avalabs) [inline x/programs/runtime/util_test.go:23]: Let's not add these and then use the CallContext instead.

File: x/programs/cmd/simulator/src/id.rs [hunk 4]
```diff
@@ -1,6 +1,6 @@
 use serde::Serialize;
 
-#[derive(Clone, Copy, Debug, PartialEq, Serialize)]
+#[derive(Clone, Copy, Debug, PartialEq, Serialize, Default)]
```
Comment (richardpringle) [inline x/programs/cmd/simulator/src/id.rs:3]: 🤔, why do you need a default here now but you didn't before?

File: x/programs/cmd/simulator/src/lib.rs [hunk 5]
```diff
@@ -91,6 +91,49 @@ pub enum Key {
...
 }
 
+#[derive(Clone, Debug, PartialEq, Default, Serialize)]
+#[serde(rename_all = "camelCase")]
+#[non_exhaustive]
+pub struct TestContext {
+    program_id: Id,
+    actor_key: Option<Key>,
+    height: u64,
+    timestamp: u64,
```
Comment (richardpringle) [inline x/programs/cmd/simulator/src/lib.rs:101]: Why not make all of these fields `pub`, this is only meant to be used in tests, and there are no properties that you need to be private

File: x/programs/cmd/simulator/src/lib.rs [hunk 6]
```diff
@@ -91,6 +91,49 @@ pub enum Key {
...
+
+    pub fn with_timestamp(mut self, timestamp: u64) -> Self {
+        self.timestamp = timestamp;
+        self
+    }
+
+    pub fn clear_actor_key(&mut self) {
+        self.actor_key = None;
+    }
+}
```
Comment (richardpringle) [inline x/programs/cmd/simulator/src/lib.rs:135]: This is overkill when you can set every property:
```suggestion
```

File: x/programs/cmd/simulator/src/lib.rs [hunk 7]
```diff
@@ -99,6 +142,8 @@ pub enum Param {
     String(String),
     Id(Id),
     Key(Key),
+    #[allow(private_interfaces)]
+    TestContext(SimulatorTestContext),
```
Comment (richardpringle) [inline x/programs/cmd/simulator/src/lib.rs:146]: Why do we have to do this?

File: x/programs/cmd/simulator/src/lib.rs [hunk 8]
```diff
@@ -188,6 +240,14 @@ where
     serializer.serialize_str(&b64.encode(text))
 }
 
+fn base64_struct_encode<S>(struc: &TestContext, serializer: S) -> Result<S::Ok, S::Error>
+where
+    S: Serializer,
+{
+    let bytes = serde_json::to_vec(struc).unwrap();
+    serializer.serialize_str(&b64.encode(bytes))
```
Comment (richardpringle) [inline x/programs/cmd/simulator/src/lib.rs:248]: What's the benefit of doing this?

File: x/programs/runtime/runtime_test.go [hunk 9]
```diff
@@ -57,3 +58,27 @@ func TestRuntimeCallProgramComplexReturn(t *testing.T) {
...
+
+	_, err = program.Call("can_change_height")
+	require.NoError(err)
+	_, err = program.CallWithHeight(1, "can_change_height")
+	require.NoError(err)
+
+	_, err = program.Call("can_change_actor")
+	require.NoError(err)
+	_, err = program.CallWithActor(codec.CreateAddress(0, ids.GenerateTestID()), "can_change_actor")
+	require.NoError(err)
```
Comment (richardpringle) [inline x/programs/runtime/runtime_test.go:83]: I think each of these should be in their own test. You should be able to instantiate the program once and then use it in each proceeding test. I've done it before but don't remember what I did.

File: x/programs/runtime/util_test.go (hunk 3 above)
Comment (richardpringle) [reply x/programs/runtime/util_test.go:23]: +1

File: x/programs/rust/examples/counter-external/src/lib.rs [hunk 10]
```diff
@@ -38,15 +38,20 @@ mod tests {
...
         let counter = simulator
             .run_step(&owner, &Step::create_program(counter_path))
             .expect("should be able to create the counter")
             .id;
         let counter = Param::Id(counter);
 
-        let params = vec![counter_external, counter.clone(), owner_key.clone()];
+        let params = vec![
+            TestContext::default()
+                .with_program_id(counter_external)
```
Comment (richardpringle) [inline x/programs/rust/examples/counter-external/src/lib.rs:50]: It would be better to do `impl From<Id> for TextContext` where all the other fields get default values.

File: x/programs/rust/examples/counter/src/lib.rs
Comment (richardpringle) [inline x/programs/rust/examples/counter/src/lib.rs:1]: If you always have to enter the program-id, then you shouldn't be able to set a default value. This is an anti-pattern

File: x/programs/rust/wasmlanche-sdk/src/program.rs [hunk 11]
```diff
@@ -25,22 +25,13 @@ pub enum ExternalCallError {
...
-    fn serialize<W: std::io::prelude::Write>(&self, writer: &mut W) -> std::io::Result<()> {
-        let Self {
-            account,
-            state_cache: _,
-        } = self;
-
-        account.serialize(writer)
-    }
-}
-
```
Comment (richardpringle) [inline x/programs/rust/wasmlanche-sdk/src/program.rs:43]: Why did you remove this?

File: x/programs/rust/wasmlanche-sdk/src/lib.rs [hunk 12]
```diff
@@ -34,24 +35,6 @@ pub struct Context<K = ()> {
...
-            action_id,
-        } = self;
-        BorshSerialize::serialize(program, writer)?;
-        BorshSerialize::serialize(actor, writer)?;
-        BorshSerialize::serialize(height, writer)?;
-        BorshSerialize::serialize(timestamp, writer)?;
-        BorshSerialize::serialize(action_id, writer)?;
-        Ok(())
-    }
-}
```
Comment (richardpringle) [inline x/programs/rust/wasmlanche-sdk/src/lib.rs:53]: Why did you remove this?

File: x/programs/test/programs/context_injection/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+            assert!(new_timestamp > timestamp)
+        }
+        None => {
+            context
+                .program
+                .state()
+                .store(StateKeys::LastTimestamp, &new_timestamp)
+                .unwrap();
+        }
+    }
```
Comment (richardpringle) [inline x/programs/test/programs/context_injection/src/lib.rs:29]: Why don't you just return the timestamp then validate that it was the one that was injected. This is overcomplicated.

File: x/programs/test/programs/context_injection/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+        Some(actor) => assert!(new_actor != actor),
+        None => {
+            context
+                .program
+                .state()
+                .store(StateKeys::LastActor, &new_actor)
+                .unwrap();
+        }
+    }
+}
```
Comment (richardpringle) [inline x/programs/test/programs/context_injection/src/lib.rs:60]: Same goes for these.

File: x/programs/test/programs/context_injection/Cargo.toml
Comment (richardpringle) [inline x/programs/test/programs/context_injection/Cargo.toml:1]: If I'm not mistaken, all the tests that are in `x/programs/tests` are meant to be tested from Go. This seems more like an e2e test that belongs in the `x/programs/cmd/simulator` directory somewhere

**Review by richardpringle — CHANGES_REQUESTED** [review]

File: x/programs/runtime/util_test.go (hunk 3 above)
Comment (iFrostizz) [reply x/programs/runtime/util_test.go:23]: Replaced fields by a `Context` instead.

File: x/programs/cmd/simulator/src/id.rs (hunk 4 above)
Comment (iFrostizz) [reply x/programs/cmd/simulator/src/id.rs:3]: That's because `TestContext` has `program_id` as a field which is of type `Id`.

File: x/programs/cmd/simulator/src/lib.rs (hunk 7 above)
Comment (iFrostizz) [reply x/programs/cmd/simulator/src/lib.rs:146]: Because we want the user space to only have access to `TestContext`, `SimulatorTestContext` is private and only meant for the simulator, because struct cannot be externally tagged. See https://github.com/ava-labs/hypersdk/pull/1076#discussion_r1651672553

File: x/programs/cmd/simulator/src/lib.rs (hunk 8 above)
Comment (iFrostizz) [reply x/programs/cmd/simulator/src/lib.rs:248]: The Go simulator reads the value from the `testContext` parameter (Type field) as a `[]byte`. Thus we need to encode this inner one as base64 because that's what `encoding/json` does on the go side. I hate it.

File: x/programs/runtime/runtime_test.go (hunk 9 above)
Comment (iFrostizz) [reply x/programs/runtime/runtime_test.go:83]: Do you mean that the test should be instantiated globally and that each test should call only one function ? That sounds like fixture-based testing

File: x/programs/rust/examples/counter-external/src/lib.rs (hunk 10 above)
Comment (iFrostizz) [reply x/programs/rust/examples/counter-external/src/lib.rs:50]: Agree with this design decision. Updated!

File: x/programs/rust/examples/counter/src/lib.rs
Comment (iFrostizz) [reply x/programs/rust/examples/counter/src/lib.rs:1]: Makes sense, we should never need to call the empty program id.

File: x/programs/rust/wasmlanche-sdk/src/program.rs (hunk 11 above)
Comment (iFrostizz) [reply x/programs/rust/wasmlanche-sdk/src/program.rs:43]: We achieve the same result with `#[borsh(skip)]`

File: x/programs/rust/wasmlanche-sdk/src/lib.rs (hunk 12 above)
Comment (iFrostizz) [reply x/programs/rust/wasmlanche-sdk/src/lib.rs:53]: Because the `BorshSerialize` attribute does the same thing

File: x/programs/test/programs/context_injection/src/lib.rs (hunk 13 above)
Comment (iFrostizz) [reply x/programs/test/programs/context_injection/src/lib.rs:29]: true... Your idea is more "atomic"

File: x/programs/test/programs/context_injection/src/lib.rs (hunk 14 above)
Comment (iFrostizz) [reply x/programs/test/programs/context_injection/src/lib.rs:60]: Changed them as well.

File: x/programs/test/programs/context_injection/Cargo.toml
Comment (iFrostizz) [reply x/programs/test/programs/context_injection/Cargo.toml:1]: Done, I moved this test in `x/programs/cmd/simulator/test/`

**Comment by iFrostizz** [comment]
All of the CI is ran because it thinks that not only programs have been ran, we might want to extend the grep range

File: x/programs/rust/wasmlanche-sdk/src/lib.rs (hunk 12 above)
Comment (richardpringle) [reply x/programs/rust/wasmlanche-sdk/src/lib.rs:75]: Can you put this back? This evaluates to 
```rust
impl<K: BorshSerialize> for Context<K> { // ...
```
But we don't need the trait bound on `K`.

File: x/programs/runtime/runtime_test.go [hunk 15]
```diff
@@ -57,3 +58,24 @@ func TestRuntimeCallProgramComplexReturn(t *testing.T) {
...
+	require.Equal(uint64(1), into[uint64](result))
+
+	result, err = program.CallWithHeight(1, "get_height")
+	require.NoError(err)
+	require.Equal(uint64(1), into[uint64](result))
+
+	result, err = program.CallWithActor(codec.CreateAddress(0, ids.GenerateTestID()), "get_actor")
+	require.NoError(err)
+	require.NotEqual(ids.Empty, into[ids.ID](result))
```
Comment (richardpringle) [inline x/programs/runtime/runtime_test.go:80]: These should be three different tests. You can use `t.Run` like I did in this PR to separate these into three different tests.

The issue is that if `CallWithTimeStamp` fails, you should still run `CallWithHeight` and `CallWithActor`. 

When anyone is debugging this, it's a lot more telling where the issue is if three different assertions fail rather than adding a patch only to get the first assertion to pass. Do you see what I mean? Unit tests are supposed to help with debugging as much as they're meant to prove that something works as expected.

File: x/programs/cmd/simulator/test/context_injection/src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+
+#[public]
+pub fn get_height(context: Context) -> u64 {
+    context.height()
+}
+
+#[public]
+pub fn get_actor(context: Context) -> Address {
+    context.actor()
+}
```
Comment (richardpringle) [inline x/programs/cmd/simulator/test/context_injection/src/lib.rs:16]: 👍

File: x/programs/cmd/simulator/test/context_injection/src/lib.rs [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+                },
+            )
+            .unwrap()
+            .result
+            .response::<Address>()
+            .unwrap();
+
+        assert_ne!(response, Address::default());
+    }
+}
```
Comment (richardpringle) [inline x/programs/cmd/simulator/test/context_injection/src/lib.rs:140]: Way better, 🥇

**Review by richardpringle — CHANGES_REQUESTED** [review]

File: x/programs/rust/wasmlanche-sdk/src/lib.rs (hunk 12 above)
Comment (iFrostizz) [reply x/programs/rust/wasmlanche-sdk/src/lib.rs:75]: Sorry ! Indeed I expanded the code and it constraint `K` to be `BorshSerialize`, this is not the case on the automatic serialization for `Program` though. I'm backing this off.

File: x/programs/runtime/runtime_test.go (hunk 15 above)
Comment (iFrostizz) [reply x/programs/runtime/runtime_test.go]: Thanks, yeah you are right that isolating unit tests makes them much easier to debug. I have used `t.Run` but it feels like the test is repeating. David's PR on default struct does not solve that because the function would panic if the default value was already set.
I have probably found a happy solution, passed a function pointer in one of the fields of the test property and executed the changing logic here.

Comment (richardpringle) [reply x/programs/runtime/runtime_test.go]: 😅 
>like I did in this PR

I forgot to provide the link

**Review by richardpringle — APPROVED** [review]
